### PR TITLE
Handle GitHub App installation deleted event

### PR DESCRIPTION
**Pull request by jhrozek**
# Summary

Adds a new HTTP handler for handling the github app. The handler fires an event
to a newly added provider installation handler which in turns talks to the
provider service that handles the deletion.

There's some additional refactoring as well as questions about provider and installation
relationships that I'll point out inline in the PR.

Fixes: #2727

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [x] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [ ] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

I tested two scenarios:
1) "claimed" installation that I created with `minder provider enroll --provider=github-app --owner=jakubtestorg`
and then deleted from the org page. This hits the codepath that deletes both the installation and the provider.
2) "unclaimed" installation that I created by navigating to `https://github.com/apps/jakub-s-local-minder-test/installations/new` where `jakub-s-local-minder-test` is my gh app name and then deleting the app from the org page. This only hits the codepath that deletes the installation.

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [ ] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [ ] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

### CODE REVIEW

File: database/query/providers.sql
Comment (jhrozek) [inline database/query/providers.sql:61]: Note that this function was unused before. Given that we already are passing UUID I hope that it should be possible to delete without passing the project as well?

@JAORMX WDYT?

File: internal/providers/service.go
Comment (jhrozek) [inline internal/providers/service.go:279]: @eleftherias since the table seems to be GH app specific and GH app installation IDs are Int64, why was the installation ID column set as TEXT?

```sql
CREATE TABLE provider_github_app_installations (
    app_installation_id TEXT PRIMARY KEY,
```

File: internal/providers/installations.go
Comment (jhrozek) [inline internal/providers/installations.go:89]: I'm not thrilled with leaking both the GH app class here and the GH app specific payload struct. I was thinking about having a callback registered on creating a GH app provider with the installation manager instead so that the installation manager would then just register class->callback map and pass on the payload as `[]byte` to the callback. But I didn't want to over-engineer without at least one review pass ...

File: internal/controlplane/handlers_githubwebhooks.go
Comment (jhrozek) [inline internal/controlplane/handlers_githubwebhooks.go:117]: There is a fair bit of code duplication between this webhook and the repo webhook. I can either:
1) do a quick refactor now to split the common code into functions
2) have the providers register a handler with the server

Ideally I'd like all github specific code out of controlplane, but maybe that's work for another day?

File: internal/providers/service.go
Comment (eleftherias) [reply internal/providers/service.go:279]: We created the tables before actually using them, and we missed this detail. I'm happy to create a migration to make it a BIGINT.

File: internal/providers/service.go
Comment (eleftherias) [reply internal/providers/service.go:280]: I think it's ok to return an error. In `DeleteProfile` we also don't treat `sql.ErrNoRows` as a special case.

File: database/query/provider_github_app_installations.sql
Comment (JAORMX) [inline database/query/provider_github_app_installations.sql:5]: Do we have an index for this?

File: internal/controlplane/handlers_githubwebhooks.go
Comment (JAORMX) [reply internal/controlplane/handlers_githubwebhooks.go:117]: Work for another day, but we should target moving it at some point

File: internal/providers/installations.go
Comment (eleftherias) [reply internal/providers/installations.go:89]: I say we skip the check on the provider class. This handler is github specific, as with our other event handlers. I think if we want to reuse them for other providers in the future, then we'll need to come up with a broad approach and apply it across the board.

File: database/query/provider_github_app_installations.sql
Comment (jhrozek) [reply database/query/provider_github_app_installations.sql:5]: we don't. I'll do that along with the int64 conversion.

File: internal/providers/installations.go
Comment (dmjb) [inline internal/providers/installations.go:1]: I see from the coverage report that there's no coverage here. Can we get some unit tests for this?

File: internal/providers/service.go
Comment (dmjb) [inline internal/providers/service.go:278]: Can we get some unit test coverage for this? It does not look like there is pre-existing coverage for this.

**Review by dmjb — CHANGES_REQUESTED** [review]

File: internal/providers/service.go
Comment (jhrozek) [reply internal/providers/service.go:278]: Yes fair points I will add coverage. I wanted to get some feedback on the approach first - maybe I should have opened a draft to not confuse people

Comment (dmjb) [reply internal/providers/service.go:278]: That's fair - just want to make sure some more test coverage is added before merge to main.

Comment (jhrozek) [reply internal/providers/service.go:278]: I opened https://github.com/stacklok/minder/pull/2885 but because there are quite a few PRs touching that logic I would prefer to unblock this PR in the meantime and work on merging the tests as we touch services.go.

**Review by eleftherias — APPROVED** [review]
Approving to unblock, since there are lots of PRs in parallel

File: internal/providers/service.go
Comment (jhrozek) [reply internal/providers/service.go:279]: let's do it in separate patch - I initially added one to the branch but as discussed elsewehere it would take too long to merge this PR. I'll send that as a separate PR and it lands when it lands.

File: internal/providers/installations.go
Comment (jhrozek) [reply internal/providers/installations.go]: I removed the provider class check (but I kept it in the message, the class might come handy in the message and it's just one set - if you prefer to nuke it altogether, I'm OK with it too)

File: database/query/provider_github_app_installations.sql
Comment (jhrozek) [reply database/query/provider_github_app_installations.sql:5]: oops, sorry, actually we do - it's a primary key so it's indexed by default (found this out while writing the conversion to int64)

File: internal/providers/installations.go
Comment (jhrozek) [reply internal/providers/installations.go:1]: added some basic coverage, thanks for reminding me!

**Comment by jhrozek** [comment]
> Approving to unblock, since there are lots of PRs in parallel

@dmjb needs to remove his changes requested, too

File: internal/providers/service.go
Comment (jhrozek) [reply internal/providers/service.go:279]: also added tests

File: database/query/providers.sql
Comment (jhrozek) [reply database/query/providers.sql:61]: also added tests

File: internal/providers/service.go
Comment (jhrozek) [reply internal/providers/service.go:279]: > also added tests

sorry wrong thread (unresolving so this one keeps being visible)

File: internal/controlplane/handlers_githubwebhooks.go
Comment (evankanderson) [inline internal/controlplane/handlers_githubwebhooks.go:152]: Do you want extract these values on 129, and be able to use `l` on 137 and beyond?

Also, a slight preference for `l := zerolog.Ctx(r.Context())...`, rather than adding a temp value whose lifetime I need to reason about.

File: internal/controlplane/handlers_githubwebhooks.go
Comment (evankanderson) [inline internal/controlplane/handlers_githubwebhooks.go:369]: The error handling / control flow here sets my teeth on edge, but I don't have a better alternative yet.

File: internal/controlplane/handlers_githubwebhooks.go
Comment (evankanderson) [inline internal/controlplane/handlers_githubwebhooks.go:388]: This feels weird... the function seems to be acting on a side-effect, but the method name is "Provider...Message", not "Send..." or "Publish..." or the like.

File: internal/providers/installations.go
Comment (evankanderson) [inline internal/providers/installations.go:77]: This feels like it might want to be a `switch` with a `default`, but this is okay for now.

File: internal/providers/installations.go
Comment (evankanderson) [inline internal/providers/installations.go:92]: This doesn't seem to return anything.  Also, the method starts with a noun; can this be `FillInstallationRemoved(msg, ...)` or the like?

Also, it feels weird to pass a `payload []byte` in here -- can we pass in a struct and have this do the marshalling?

Alternatively, I could see this growing up to actually create the message and set most of the fields on it.

File: internal/providers/service.go
Comment (evankanderson) [reply internal/providers/service.go:279]: I used `TEXT` despite it being GitHub-specific because I didn't want to attach too firmly to the implementation detail that these are int64 (for example, if we needed something different for GH Enterprise), but I'm happy to switch to int64 if we prefer.

I do think that there are a bunch of details to GitHub Apps that are probably unique, so using a column type that matches GitHub Apps closely doesn't really bother me here.

File: internal/providers/service.go
Comment (evankanderson) [inline internal/providers/service.go:295]: I think this is okay for now, but we should make sure that the flow here and in the `DeleteProvider` gRPC endpoint end up staying the same.

File: internal/controlplane/handlers_githubwebhooks.go
Comment (evankanderson) [reply internal/controlplane/handlers_githubwebhooks.go:369]: To be specific: it seems like `errNotHandled` is a success-error, which then gets consumed and transformed in some way by `handleParseError`, and the result of that interacts with GitHub's webhook retry-on-error logic.

**Review by evankanderson — APPROVED** [review]
No blocking comments, but the most serious is the removal of needing to know the project context from `DeleteProvider`, simply due to our experience of previous security issues there.  I might almost be tempted to create a `DeleteProviderUnconditional` or `DeleteProviderWithoutProject` to discourage people from using the non-project form where possible.

**Review by dmjb — APPROVED** [review]

**Comment by jhrozek** [comment]
@eleftherias may I ask for another approve? I just rebased atop the current main. Sorry for the churn, I will try to split my PRs more granularly in the future.

Also @evankanderson you raised really good points and I will fix them, but let's get this merged so we avoid more back-and-forth at this moment. Thank you for the review, I was glad someone called me out on the project-less delete!

File: database/query/provider_github_app_installations.sql
Comment (eleftherias) [inline database/query/provider_github_app_installations.sql:4]: non-blocking: A better name for this would be `GetInstallationByInstallationID`, since app ID is something else.

**Review by eleftherias — APPROVED** [review]
Added one more comment for the follow up PRs

File: database/query/provider_github_app_installations.sql
Comment (jhrozek) [reply database/query/provider_github_app_installations.sql:4]: thank you I will take this into account

**Pull request merged**
